[fetch] Implement `Headers.prototype.getSetCookie`.

andreubotella opened the pull request:
#### 77fd8deab6cb8bda8d3e265dc91ab65685fdc84d
<pre>
[fetch] Implement `Headers.prototype.getSetCookie`.
<a href="https://bugs.webkit.org/show_bug.cgi?id=251194">https://bugs.webkit.org/show_bug.cgi?id=251194</a>

Reviewed by Youenn Fablet.

This change implements the `getSetCookie` method of the fetch spec&apos;s
`Headers` interface. This change allows storing `Set-Cookie` headers as
a list of values, and retrieving the separate values, rather than their
combined representation, which is ambiguous. This change also makes it
so iterating through `Headers` objects will yield `Set-Cookie` headers
separately rather than combined, which is part of the same change in the
fetch spec.

Given that `Set-Cookie` is a forbidden header name in both requests and
responses, this will not be useful for any uses of `Headers` related to
fetch. It does allows some non-fetch use cases, however, and enables
further compatibility with server-side runtimes that do allow this
header.

Rather than refactoring `HTTPHeaderMap` to not combine (all) headers,
this implementation makes use of the fact that `Set-Cookie` is forbidden
for both requests and responses, to store the `Set-Cookie` values in a
vector of strings directly in `FetchHeaders` instead. This needed such
header name to be special-cased in almost every operation of this class.
And in order to have `Set-Cookie` headers properly appear in their right
order in the pair iterator, `m_keys` has been changed to be a pair where
the second element is an index into the `Set-Cookie` values vector,
which is ignored for other header names.

Fetch spec PR: <a href="https://github.com/whatwg/fetch/pull/1346">https://github.com/whatwg/fetch/pull/1346</a>

This commit also imports the corresponding WPT tests from
<a href="https://github.com/web-platform-tests/wpt/pull/31442">https://github.com/web-platform-tests/wpt/pull/31442</a>, and adds some
additional tests.

* Source/WebCore/Modules/fetch/FetchHeaders.cpp:
(WebCore::appendToHeaderMap):
(WebCore::fillHeaderMap):
(WebCore::FetchHeaders::create):
(WebCore::FetchHeaders::fill):
(WebCore::FetchHeaders::append):
(WebCore::FetchHeaders::remove):
(WebCore::FetchHeaders::get const):
(WebCore::FetchHeaders::getSetCookie const):
(WebCore::FetchHeaders::has const):
(WebCore::FetchHeaders::set):
(WebCore::FetchHeaders::Iterator::next):
* Source/WebCore/Modules/fetch/FetchHeaders.h:
(WebCore::FetchHeaders::create):
(WebCore::FetchHeaders::fastGet const):
(WebCore::FetchHeaders::fastHas const):
(WebCore::FetchHeaders::fastSet):
(WebCore::FetchHeaders::FetchHeaders):
* Source/WebCore/Modules/fetch/FetchHeaders.idl:
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::httpHeaderField const):

Canonical link: <a href="https://commits.webkit.org/260533@main">https://commits.webkit.org/260533@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c6b517708ea0783b6584fc2482d00defd414b8be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108634 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17735 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41488 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117744 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/117945 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112517 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19186 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9012 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100866 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114401 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14384 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97618 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42354 "Found 3 new test failures: webgl/2.0.y/conformance/glsl/bugs/character-set.html, webgl/2.0.y/conformance/glsl/misc/fragcolor-fragdata-invariant.html, webgl/2.0.y/conformance2/renderbuffers/multisampled-depth-renderbuffer-initialization.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96360 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29256 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10543 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30606 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11301 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7519 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16692 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50202 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7278 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12889 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.cpp:
```diff
@@ -57,7 +57,7 @@ static ExceptionOr<bool> canWriteHeader(const String& name, const String& value,
 }
 
-static ExceptionOr<void> appendToHeaderMap(const String& name, const String& value, HTTPHeaderMap& headers, FetchHeaders::Guard guard)
+static ExceptionOr<void> appendToHeaderMap(const String& name, const String& value, HTTPHeaderMap& headers, Vector<String>& setCookieValues, FetchHeaders::Guard guard)
```
It might be handy to introduce a struct InternalHeaders {
HTTPHeaderMap;
Vector<String>&;
};

andreubotella replied:
Is the idea to use this struct on `FetchHeaders::internalHeaders()` and `setInternalHeaders()` as well? As far as I can tell, those methods are only used for fetch requests and responses, and shouldn't need to get or set the `Set-Cookie` values.

youennf replied:
Yeah, but this seems good enough the way you did it.

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.cpp:
```diff
@@ -229,15 +269,20 @@ std::optional<KeyValuePair<String, String>> FetchHeaders::Iterator::next()
-        auto value = m_headers->m_headers.get(key);
+        auto pair = m_keys[m_currentIndex++];
+        String key(WTFMove(pair.first));
+        auto value = (key == "set-cookie"_str) ? m_headers->m_setCookieValues[pair.second] : m_headers->m_headers.get(key);
```
We probably do not need to have a Vector<String, size_t> here.
We could simply keep a m_setCoookieCounter in the Iterator and when String is set_cookie, increment it/

andreubotella replied:
That wouldn't work when the headers are modified in the middle of the iteration. But if we also store the current size of `m_setCookieValues` in the iterator, this could be worked around in the invalidation recovery code in `next()`. I'll give that a try.

andreubotella replied:
I changed this to use a `m_setCookieIndex` rather than pairs. As it turns out, keeping the previous size of `m_setCookieValues` wasn't needed. Please let me know if this needs some more documentation.

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.h:
```diff
@@ -52,22 +52,35 @@ class FetchHeaders : public RefCounted<FetchHeaders> {
+    {
+        ASSERT(name != HTTPHeaderName::SetCookie);
+        m_headers.set(name, value);
+    }
```
Since they are multilines, we can keep them in the header file but after the class declaration:
inline String FetchHeaders::fastGet(..._ {
}

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.cpp:
```diff
@@ -68,7 +68,11 @@ static ExceptionOr<void> appendToHeaderMap(const String& name, const String& val
         return { };
-    headers.set(name, combinedValue);
+
+    if (equalIgnoringASCIICase(name, "Set-Cookie"_str))
```
We might want to special case set-cookie a bit earlier.
The computation of the combined value is done using headers and not setCookieValues for Set-Cookie.
I do not think we need to combine values for SetCookie validation.

youennf reviewed (commented):
Some suggestions below.
This patch needs tests, ideally WPT tests.

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.cpp:
```diff
@@ -144,14 +150,19 @@ ExceptionOr<void> FetchHeaders::fill(const FetchHeaders& otherHeaders)
             return result.releaseException();
     }
+    for (auto& setCookieValue : otherHeaders.m_setCookieValues) {
+        auto result = appendToHeaderMap("Set-Cookie"_str, setCookieValue, m_headers, m_setCookieValues, m_guard);
```
I would directly introduce a appendSetCookie function.
appendToHeaderMap would also call appendSetCookie very early as well.

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.cpp:
```diff
@@ -228,16 +268,46 @@ void FetchHeaders::filterAndFill(const HTTPHeaderMap& headers, Guard guard)
-        auto value = m_headers->m_headers.get(key);
+        auto key = m_keys[m_currentIndex];
+        String value;
+        if (key == "set-cookie"_str) {
```
It is a bit sad we are doing this check, this is not really efficient.
Can we instead have m_keys storing SetCookieOrOtherName? Or maybe consider the null string is Set-Cookie?

andreubotella replied:
If we treat the null string, or some other value, as `Set-Cookie`, we would have to use a custom comparison function when sorting and finding the index of `Set-Cookie`. But I was wondering if there is some way to optimize this check with `AtomString` instead.

youennf replied:
The custom function would basically call codePointCompareLessThan except if a or b is null in which case it would call codePointCompareLessThan with a or b being set to "set-cookie".

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.cpp:
```diff
@@ -228,16 +268,46 @@ void FetchHeaders::filterAndFill(const HTTPHeaderMap& headers, Guard guard)
+        } else {
+            m_currentIndex++;
+            value = m_headers->m_headers.get(key);
+        }
```
Do we have enough test coverage here?
Say we have 3 Set-cookie headers and are in the middle of Set-Cookie iterations.
We then remove Set-Cookie and append 3 new Set-Cookie headers.

andreubotella replied:
There is a test that makes sure that adding an earlier header while iterating in the middle of `Set-Cookie`, and that adding a `Set-Cookie` header after iterating past them, behave as expected. But I can add that extra test.

youennf replied:
Please make sure the tests cover both code paths.

youennf commented on LayoutTests/imported/w3c/resources/import-expectations.json:
```diff
@@ -231,6 +231,7 @@
     "web-platform-tests/fetch": "import",
     "web-platform-tests/fetch/api": "import",
     "web-platform-tests/fetch/api/cors": "import",
+    "web-platform-tests/fetch/api/headers": "import",
```
not needed

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.cpp:
```diff
@@ -181,13 +207,35 @@ ExceptionOr<String> FetchHeaders::get(const String& name) const
+
+    if (equalIgnoringASCIICase(name, "Set-Cookie"_str)) {
+        if (m_setCookieValues.isEmpty())
+            return String();
```
We probably do not need to optimize this case.

andreubotella replied:
This is just about the size-0 case, right? The size-1 optimization can stay.

andreubotella replied:
Oh, I hadn't realized that returning `String()` on the size-0 case isn't an optimization, since `get()` must return `null`. The size-1 case is an optimization, but not apparently one that would achieve anything, since it seems like `StringBuilder` directly uses the passed string in this case.

I'm therefore keeping the size-0 case and removing the size-1 case.

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.cpp:
```diff
@@ -201,7 +249,11 @@ ExceptionOr<void> FetchHeaders::set(const String& name, const String& value)
-    m_headers.set(name, normalizedValue);
+    if (equalIgnoringASCIICase(name, "Set-Cookie"_str)) {
+        m_setCookieValues.clear();
+        m_setCookieValues.append(value);
```
Shouldn't it be normalizedValue? Do we have tests for this?

andreubotella replied:
You're right. I added some tests.

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.cpp:
```diff
@@ -225,19 +277,62 @@ void FetchHeaders::filterAndFill(const HTTPHeaderMap& headers, Guard guard)
     }
 }
 
+inline bool compareIteratorKeys(const String& a, const String& b)
```
we usually use static here instead of inline

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.cpp:
```diff
@@ -225,19 +277,62 @@ void FetchHeaders::filterAndFill(const HTTPHeaderMap& headers, Guard guard)
+{
+    // null in the iterator's m_keys represents Set-Cookie.
+    if (a.isNull() && b.isNull())
+        return false;
```
Do we need this first if?

andreubotella replied:
No. I added it as a performance optimization, but I guess it's not needed, considering there's only a single `set-cookie` header in `m_keys`.

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.cpp:
```diff
@@ -225,19 +277,62 @@ void FetchHeaders::filterAndFill(const HTTPHeaderMap& headers, Guard guard)
-        auto value = m_headers->m_headers.get(key);
+        auto key = m_keys[m_currentIndex];
+        String value;
+        if (key.isNull()) {
```
We usually tend to use return  and  so on to reduce if / else.
That would mean something like:
```
if (key.isNull()) {
    if (m_setCookieIndex < m_headers->m_setCookieValues.size())
        return KeyValuePair<String, String> { ..., ...};
    m_currentIndex++;
    continue;
}
m_currentIndex++;
...

```

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.cpp:
```diff
@@ -57,9 +57,25 @@ static ExceptionOr<bool> canWriteHeader(const String& name, const String& value,
+static ExceptionOr<void> appendToHeaderMap(const String& name, const String& value, HTTPHeaderMap& headers, Vector<String>& setCookieValues, FetchHeaders::Guard guard)
 {
     String normalizedValue = stripLeadingAndTrailingHTTPSpaces(value);
+    if (equalIgnoringASCIICase(name, "Set-Cookie"_str))
```
Why camel-cased `"Set-Cookie"_str`? Wouldn't it be better to use lower cased ASCIILiteral `"set-cookie"_s`?

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.cpp:
```diff
@@ -181,13 +207,33 @@ ExceptionOr<String> FetchHeaders::get(const String& name) const
     return m_headers.get(name);
 }
 
+Vector<String> FetchHeaders::getSetCookie() const
```
Should probably be const Vector<String>&

andreubotella replied:
Fixed.

youennf reviewed (approved):
LGTM

andreubotella committed:
[fetch] Implement `Headers.prototype.getSetCookie`.
https://bugs.webkit.org/show_bug.cgi?id=251194

Reviewed by Youenn Fablet.

This change implements the `getSetCookie` method of the fetch spec's
`Headers` interface. This change allows storing `Set-Cookie` headers as
a list of values, and retrieving the separate values, rather than their
combined representation, which is ambiguous. This change also makes it
so iterating through `Headers` objects will yield `Set-Cookie` headers
separately rather than combined, which is part of the same change in the
fetch spec.

Given that `Set-Cookie` is a forbidden header name in both requests and
responses, this will not be useful for any uses of `Headers` related to
fetch. It does allows some non-fetch use cases, however, and enables
further compatibility with server-side runtimes that do allow this
header.

Rather than refactoring `HTTPHeaderMap` to not combine (all) headers,
this implementation makes use of the fact that `Set-Cookie` is forbidden
for both requests and responses, to store the `Set-Cookie` values in a
vector of strings directly in `FetchHeaders` instead. This needed such
header name to be special-cased in almost every operation of this class.
And in order to have `Set-Cookie` headers properly appear in their right
order in the pair iterator, `m_keys` has been changed to be a pair where
the second element is an index into the `Set-Cookie` values vector,
which is ignored for other header names.

Fetch spec PR: https://github.com/whatwg/fetch/pull/1346

This commit also imports the corresponding WPT tests from
https://github.com/web-platform-tests/wpt/pull/31442, and adds some
additional tests.

* Source/WebCore/Modules/fetch/FetchHeaders.cpp:
(WebCore::appendToHeaderMap):
(WebCore::fillHeaderMap):
(WebCore::FetchHeaders::create):
(WebCore::FetchHeaders::fill):
(WebCore::FetchHeaders::append):
(WebCore::FetchHeaders::remove):
(WebCore::FetchHeaders::get const):
(WebCore::FetchHeaders::getSetCookie const):
(WebCore::FetchHeaders::has const):
(WebCore::FetchHeaders::set):
(WebCore::FetchHeaders::Iterator::next):
* Source/WebCore/Modules/fetch/FetchHeaders.h:
(WebCore::FetchHeaders::create):
(WebCore::FetchHeaders::fastGet const):
(WebCore::FetchHeaders::fastHas const):
(WebCore::FetchHeaders::fastSet):
(WebCore::FetchHeaders::FetchHeaders):
* Source/WebCore/Modules/fetch/FetchHeaders.idl:
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::httpHeaderField const):

Canonical link: https://commits.webkit.org/260533@main
